File: VDM/Assets/Scripts/Runtime/Events/Models/EventTypes.cs
namespace VDM.Runtime.Events.Models
{
    /// <summary>
    /// Centralized event type constants aligned with backend event system.
    /// Provides consistent event type identifiers across frontend and backend.
    /// </summary>
    public static class EventTypes
    {
        // System Events
        public const string SystemInitialized = "system.initialized";
        public const string SystemShutdown = "system.shutdown";
        public const string GameStarted = "game.started";
        public const string GameStopped = "game.stopped";
        public const string GameSaved = "game.saved";
        public const string GameLoaded = "game.loaded";
        
        // Character Events
        public const string CharacterCreated = "character.created";
        public const string CharacterUpdated = "character.updated";
        public const string CharacterDeleted = "character.deleted";
        public const string CharacterLeveledUp = "character.leveled_up";
        
        // Combat Events
        public const string CombatStarted = "combat.started";
        public const string CombatEnded = "combat.ended";
        public const string CombatTurnStarted = "combat.turn_started";
        public const string CombatTurnEnded = "combat.turn_ended";
        
        // Quest Events
        public const string QuestCreated = "quest.created";
        public const string QuestUpdated = "quest.updated";
        public const string QuestCompleted = "quest.completed";
        public const string QuestFailed = "quest.failed";
        
        // Faction Events
        public const string FactionCreated = "faction.created";
        public const string FactionUpdated = "faction.updated";
        public const string FactionRelationshipChanged = "faction.relationship_changed";
        
        // NPC Events
        public const string NpcCreated = "npc.created";
        public const string NpcUpdated = "npc.updated";
        public const string NpcDeleted = "npc.deleted";
        public const string NpcInteraction = "npc.interaction";
        
        // Dialogue Events
        public const string DialogueStarted = "dialogue.started";
        public const string DialogueEnded = "dialogue.ended";
        public const string DialogueMessageSent = "dialogue.message_sent";
        
        // Time Events
        public const string TimeAdvanced = "time.advanced";
        public const string TimeScaleChanged = "time.scale_changed";
        
        // World Events
        public const string WorldGenerated = "world.generated";
        public const string WorldStateChanged = "world.state_changed";
        public const string RegionUpdated = "region.updated";
        
        // Data Events
        public const string DataLoaded = "data.loaded";
        public const string DataSaved = "data.saved";
        public const string DataValidated = "data.validated";
        public const string DataError = "data.error";
        
        // UI Events
        public const string UiPanelOpened = "ui.panel_opened";
        public const string UiPanelClosed = "ui.panel_closed";
        public const string UiButtonClicked = "ui.button_clicked";
        public const string UiInputChanged = "ui.input_changed";
        
        // Memory Events
        public const string MemoryCreated = "memory.created";
        public const string MemoryUpdated = "memory.updated";
        public const string MemoryDecayed = "memory.decayed";
        
        // Custom Events
        public const string Custom = "custom";
    }
} 

File: backend/infrastructure/events/events/event_types.py
"""
Event Types and Classes for Visual DM
------------------------------------
This module provides backward compatibility for old event type imports.
For code, please import directly from canonical_events.py.

The event system is the backbone of Visual DM's event-driven architecture.
"""

from enum import Enum, auto
from typing import Dict, Any, Optional, List, Union
from uuid import UUID

from backend.infrastructure.events.core.event_base import EventBase
# from backend.infrastructure.events.events.canonical_events import (
#     # Import all canonical events for type assignment
#     MemoryCreated, MemoryReinforced, MemoryDecayed, MemoryDeleted, MemoryRecalled,
#     RumorCreated, RumorSpread, RumorMutated,
#     MotifChanged,
#     WorldStateChanged,
#     RelationshipChanged, RelationshipCreated, RelationshipUpdated, RelationshipRemoved,
#     CharacterCreated, CharacterUpdated, CharacterDeleted, CharacterLeveledUp,
#     MoodChanged,
#     GoalCreated, GoalCompleted, GoalFailed, GoalAbandoned, GoalProgressUpdated,
#     PopulationChanged,
#     POIStateChanged,
#     FactionChanged,
#     QuestUpdated,
#     CombatEvent, CombatStarted, CombatEnded,
#     TimeAdvanced,
#     StorageEvent, GameSaved, GameLoaded,
#     EventLogged,
# )

class EventType(Enum):
    """General event type enumeration"""
    # Core system events
    SYSTEM_LOG = "system.log"
    SYSTEM_ERROR = "system.error"
    SYSTEM_DEBUG = "system.debug"
    SYSTEM_STARTUP = "system.startup"
    SYSTEM_SHUTDOWN = "system.shutdown"
    
    # Memory events
    MEMORY_CREATED = "memory.created"
    MEMORY_REINFORCED = "memory.reinforced"
    MEMORY_DECAYED = "memory.decayed"
    MEMORY_DELETED = "memory.deleted"
    MEMORY_RECALLED = "memory.recalled"
    
    # Rumor events
    RUMOR_CREATED = "rumor.created"
    RUMOR_SPREAD = "rumor.spread"
    RUMOR_MUTATED = "rumor.mutated"
    
    # Character events
    CHARACTER_CREATED = "character.created"
    CHARACTER_UPDATED = "character.updated"
    CHARACTER_DELETED = "character.deleted"
    CHARACTER_LEVELED_UP = "character.leveled_up"
    CHARACTER_MOOD_CHANGED = "character.mood_changed"
    
    # Quest events
    QUEST_UPDATED = "quest.updated"
    
    # Combat events
    COMBAT_STARTED = "combat.started"
    COMBAT_ENDED = "combat.ended"
    COMBAT_TURN = "combat.turn"
    COMBAT_ACTION = "combat.action"
    
    # World events
    WORLD_STATE_CHANGED = "world_state.changed"
    TIME_ADVANCED = "time.advanced"
    
    # Storage events
    STORAGE_SAVED = "storage.saved"
    STORAGE_LOADED = "storage.loaded"
    
    # Arc events (for the arc system we're implementing)
    ARC_CREATED = "arc.created"
    ARC_UPDATED = "arc.updated"
    ARC_COMPLETED = "arc.completed"
    ARC_FAILED = "arc.failed"
    ARC_STEP_COMPLETED = "arc.step.completed"

# System Events
class SystemEventType(Enum):
    """System event types"""
    LOG = "log"
    ERROR = "error"
    DEBUG = "debug"
    STARTUP = "startup"
    SHUTDOWN = "shutdown"

class SystemEvent(EventBase):
    """Legacy system event class - use specific event classes from canonical_events instead"""
    system_type: SystemEventType
    message: str
    details: Optional[Dict[str, Any]] = None

# Memory Events
class MemoryEventType(Enum):
    """Memory event types"""
    CREATED = "created"
    REINFORCED = "reinforced"
    DECAYED = "decayed"
    DELETED = "deleted"
    RECALLED = "recalled"

class MemoryEvent(EventBase):
    """Legacy memory event class - use specific event classes from canonical_events instead"""
    memory_type: MemoryEventType
    entity_id: Union[str, UUID]
    memory_id: Union[str, UUID]
    content: Dict[str, Any]
    importance: Optional[float] = None

# Rumor Events
class RumorEventType(Enum):
    """Rumor event types"""
    CREATED = "created"
    SPREAD = "spread"
    MUTATED = "mutated"

class RumorEvent(EventBase):
    """Legacy rumor event class - use specific event classes from canonical_events instead"""
    rumor_type: RumorEventType
    rumor_id: Union[str, UUID]
    source_id: Optional[Union[str, UUID]] = None
    target_id: Optional[Union[str, UUID]] = None
    content: Dict[str, Any]
    spread_factor: Optional[float] = None

# Motif Events
class MotifEventType(Enum):
    """Motif event types"""
    CHANGED = "changed"

class MotifEvent(EventBase):
    """Legacy motif event class - use specific event classes from canonical_events instead"""
    motif_type: MotifEventType
    motif_id: str
    old_value: Optional[float] = None
    new_value: float
    location_id: Optional[Union[str, UUID]] = None

# Population Events
class PopulationEventType(Enum):
    """Population event types"""
    CHANGED = "changed"

class PopulationEvent(EventBase):
    """Legacy population event class - use specific event classes from canonical_events instead"""
    population_type: PopulationEventType
    location_id: Union[str, UUID]
    old_count: Optional[int] = None
    new_count: int
    change_reason: Optional[str] = None

# POI Events
class POIEventType(Enum):
    """POI event types"""
    STATE_CHANGED = "state_changed"

class POIEvent(EventBase):
    """Legacy POI event class - use specific event classes from canonical_events instead"""
    poi_type: POIEventType
    poi_id: Union[str, UUID]
    old_state: Optional[Dict[str, Any]] = None
    new_state: Dict[str, Any]

# Faction Events
class FactionEventType(Enum):
    """Faction event types"""
    CHANGED = "changed"

class FactionEvent(EventBase):
    """Legacy faction event class - use specific event classes from canonical_events instead"""
    faction_type: FactionEventType
    faction_id: Union[str, UUID]
    changes: Dict[str, Any]

# Quest Events
class QuestEventType(Enum):
    """Quest event types"""
    UPDATED = "updated"

class QuestEvent(EventBase):
    """Legacy quest event class - use specific event classes from canonical_events instead"""
    quest_type: QuestEventType
    quest_id: Union[str, UUID]
    update_type: str
    changes: Dict[str, Any]

# Combat Events
class CombatEventType(Enum):
    """Combat event types"""
    STARTED = "started"
    ENDED = "ended"
    TURN = "turn"
    ACTION = "action"
    DAMAGE = "damage"
    EFFECT = "effect"

class CombatEvent(EventBase):
    """Legacy combat event class - use specific event classes from canonical_events instead"""
    combat_type: CombatEventType
    combat_id: Union[str, UUID]
    participants: List[Union[str, UUID]]
    details: Dict[str, Any]

# Time Events
class TimeEventType(Enum):
    """Time event types"""
    ADVANCED = "advanced"

class TimeEvent(EventBase):
    """Legacy time event class - use specific event classes from canonical_events instead"""
    time_type: TimeEventType
    old_time: Optional[Dict[str, Any]] = None
    new_time: Dict[str, Any]
    units_passed: Optional[int] = None

# Relationship Events
class RelationshipEventType(Enum):
    """Relationship event types"""
    CREATED = "created"
    UPDATED = "updated"
    REMOVED = "removed"

class RelationshipEvent(EventBase):
    """Legacy relationship event class - use specific event classes from canonical_events instead"""
    relationship_type: RelationshipEventType
    source_id: Union[str, UUID]
    target_id: Union[str, UUID]
    old_value: Optional[float] = None
    new_value: Optional[float] = None
    relationship_kind: Optional[str] = None

# Storage Events
class StorageEventType(Enum):
    """Storage event types"""
    SAVED = "saved"
    LOADED = "loaded"

class StorageEvent(EventBase):
    """Legacy storage event class - use specific event classes from canonical_events instead"""
    storage_type: StorageEventType
    file_path: Optional[str] = None
    data_size: Optional[int] = None
    details: Optional[Dict[str, Any]] = None

# World State Events
class WorldStateEventType(Enum):
    """World state event types"""
    CHANGED = "changed"

class WorldStateEvent(EventBase):
    """Legacy world state event class - use specific event classes from canonical_events instead"""
    world_state_type: WorldStateEventType
    changes: Dict[str, Any]
    affected_entities: Optional[List[Union[str, UUID]]] = None